MarkdownLimit Proposal

dseynhae opened the pull request:
# Existing Sample Changes
## Description

Cleaned MarkdownFile with MarkdownLint:
* Increase probability that features (e.g. code highlighting) will work with customer's MD reader.
* Remove format anomalies due to accidental indentation.
* Consistent look-feel for all different test cases.
Submitting this to see if we can encourage all people to clean Markdown files they are touching...

Fixes Issue# 

## External Dependencies

List any external dependencies created as a result of this change.

## Type of change

Please delete options that are not relevant. Add a 'X' to the one that is applicable. 

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Implement fixes for ONSAM Jiras

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Command Line
- [ ] oneapi-cli
- [ ] Visual Studio
- [ ] Eclipse IDE
- [ ] VSCode
- [ ] When compiling the compliler flag "-Wall -Wformat-security -Werror=format-security" was used

**_Delete this line and everything below if this is not a PR for a new code sample_**

**_Delete this line and all above it if this PR is for a new code sample_**
# Adding a New Sample(s)

## Description

Please include a description of the sample

## Checklist
Administrative
- [ ] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): <insert Name Here>
- [ ] If you have any new dependencies/binaries, inform the oneAPI Code Samples Project Manager: @JoeOster

Code Development
- [ ] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [ ] Adhere to readme template 
- [ ] Enforce format via clang-format config file
- [ ] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [ ] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [ ] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see @JoeOster for assistance)
- [ ] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [ ] Bandit Scans (Python only)
- [ ] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Joe Oster(@JoeOster)
- [ ] Tested using Dev Cloud when applicable

dseynhae commented:
@mdtucker: This should show up on your radar now?
I think that protected branches are probably a better fit for ensuring that nothing goes wrong with the flow...

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md:
```diff
@@ -37,114 +43,127 @@ Third party program Licenses can be found here: [third-party-programs.txt](https
+* Download a sample using the extension **Code Sample Browser for Intel oneAPI Toolkits**.
+* Configure the oneAPI environment with the extension **Environment Configurator for Intel oneAPI Toolkits**.
+* Open a Terminal in VS Code (**Terminal>New Terminal**).
+* Run the sample in the VS Code terminal using the instructions below.
```
This seems to render the same way - why the change?

dseynhae replied:
This is a consistency rule, both `-` and `*` can be used for bullets, but to avoid confusion, the linter will flag any mixture of those styles. Which style you pick is a personal preference, but the recommendation is there because the raw text should also be consistent...

mdbtucker commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md:
````diff
@@ -37,114 +43,127 @@ Third party program Licenses can be found here: [third-party-programs.txt](https
 1. Install the design into a directory `build` from the design directory by running `cmake`:
 
-   ```
+   ```bash
````
From what I understand, the purpose of tags in your text blocks is to enable syntax highlighting.  However, the highlighting depends on the Markdown renderer, and in this case there's actually nothing to highlight, so I'm not sure these tags are adding any value.

dseynhae replied:
It's like turning on your blinker when driving. You might think it doesn't matter, but if you just always do it, you won't forget when it actually does matter.

And you would be amazed what the automatic highlighter comes up with when you don't specify it...

mdbtucker reviewed (commented):
I think we should be consistent between code samples, and I'm not sure this is worth doing to all the code samples, especially given the fact that the linting tool seems to have made some mistakes (with moving the * characters) that would have to be manually corrected.  I think the effort outweighs the minimal (at best) benefits.

But I'm open to being convinced if you feel you have a compelling argument for doing this?

dseynhae commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md:
```diff
@@ -1,24 +1,29 @@
 | Optimized for                     | Description
 ---                                 |---
-| OS                                | Linux* Ubuntu* 18.04/20.04, RHEL*/CentOS* 8, SUSE* 15; Windows* 10
+| OS                                | Linux *Ubuntu* 18.04/20.04, RHEL*/CentOS* 8, SUSE *15; Windows* 10
```
Unfortunately, you might want it to be a trademark, but it is interpreted as markdown...

I can resolve this in two ways:
* Use the actual ™ symbol.
* Escape the `*` symbol: `\*`

Note that the problems are NOT fixed automatically. They can be, but I fixed every single problem by hand... I'll wait for your preference to fix the trademark indication (which was NOT working before, it was actually *emphasizing* the enclosed text.

dseynhae commented on DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/README.md:
```diff
@@ -1,24 +1,29 @@
 | Time to complete                  | 1 hour
 
 ## Purpose
+
```
There are different MD rendering engines, and they determine the style for heading rendering.
Some of those styles don't look good without the "surrounding" empty lines, so the guideline is to always use empty lines around the headers, to accommodate all possible rendering styles.

mdbtucker commented:
You make a lot of good points Dirk, and I thank you for bringing this up.  I'm going to 'kick the can' on this one and defer to @tomlenth, who owns the 'README' files.

Tom, please take a look at this review and Dirk's comments.  I think if we are going to address these issues, we should be consistent and address them for all code samples, which seems like a medium amount of work.  Or I guess we could just start enforcing new rules (such as running a linter on all README files) for all new samples and any time a README is changed.  Or we could decide that, correct though Dirk is on many of his points, it's not worth the effort.  I leave the decision to you.

dseynhae commented:
@tomlenth, @mdbtucker :
Just putting a bit of context in place here: I work in the "Strategic Planning" team for the HLD (High Level Design) team. I'm supposed to be the "UX" architect (User Experience).

As such, I'm a firm believer that the first impressions must be impecible. And those first impressions start with searching for info, our information and download sites, and then of course our "documentation".

For the OneAPI examples, that documentation is centered in the Markdown files. And when customers clone the repo, those Markdown files will no longer be rendered by GitHub. Personally, I render Markdown in VS Code, and I have the popular [Markdown Viewer](https://chrome.google.com/webstore/detail/markdown-viewer/ckkdlimhmcjmikdlpkmbgfkaikojcbjk?hl=en) for Chrome (also available for Firefox and Edge). But you just don't know what the customer will throw at it... So my comments are really geared towards optimal viewing experiences, independent from how you render it.

In my experience, a lot of the rules in the [MarkdownLint](https://github.com/DavidAnson/markdownlint) application are designed exactly towards that goal...

It also cleans up common mistakes, such as unintentional indentation (which can have an effect on how the Markdown is rendered, or wrong indentation (too many levels), wrong header hierarchy (some Markdown rendering tools extract a TOC from the Markdown code),...

In this particular case, there wasn't really an "error". An emphasis markdown is neutralized when surrounded by spaces...  But a common error is:

```markdown
I want to * emphasize * this...
```

To avoid this, `Markdownlint` will detect those as a potential problem.
But in our case, there is only a potential problem:

```markdown
Linux* Ubuntu* 18.04/20.04, RHEL*/CentOS* 8, SUSE* 15; Windows* 10
```

1. The intention MIGHT have been to emphasize `*Ubuntu*`.
2. Not all renderers are going to implement `Linux*`.

So the best thing to do, is to make clear that the `*` is NOT a Markdown metacharacter:

```markdown
Linux\* Ubuntu\* 18.04/20.04, RHEL\*/CentOS\* 8, SUSE\* 15; Windows\* 10
```

It is actually amazing how many ambiguous situations the `MarkdownLint` application prevents...

tomlenth commented:
Hi Dirk, we are kindred fellows as I'm the DX architect for oneAPI.  Getting the readme files as consistent as they are has been a challenge given the many authors and legacy samples.  This is just one small slice of the oneAPI pie that our group tries to influence.  I've used the linter with mixed success, and also worked through some of the templates keeping mind that, as you note, the material will be viewed differently depending on the users environment. I use VS Code as well, but often skip the rendered view since many of our users will view the raw file (VIM, notepad, emacs) .  We've tried to ensure the raw file is as readable as possible too.  That said, I'm not sure how much the different authors have deviated since our initial reviews and there are definitely a few that completely slipped through the cracks.  We have a new resource on our team coming up to speed who has been looking a lot at the consistency problems.  We're looking at getting a much better overall samples strategy together for 2022, including getting our Info Dev owners into the review cycles to ensure our sample readmes are consistent with the much larger realm of product documentation.  It may be of value for the three of us to meet offline. 

As for the use of the asterisks, it's the bane of my existence.  Through many user studies, we see users question what it's for.  Corporate mandates third party attribution using the asterisks.  It's not meant to provide any formatting and my VS Code extension renders it ok, without applying emphasis to the content. If it is, than indeed that is a problem and it sounds like different extensions are rendering it differently.

dseynhae commented:
@tomlenth 
Very happy to see that we’re planning/trying to address this.

The asterisk problem is easily addressed by escaping it as a metacharacter, if it is not meant to be one.

I know that a lot of MD rendering solutions try to be clever, and determine if the asterisk is supposed to be a metacharacter or not. But that in itself leads to problems!

Take:

```
This is emphasized text
```

When double-clicking `emphasized`  in some online MD editors, it will select the space before or after the word. When then using the GUI to emphasize, the underlying MarkDown becomes:

```
This is* emphasized* text
```

And that WILL be rendered differently, in the applications that try to be clever (in our situation, we would have trademarked `is` and `emphasized`.

This brings me to my last point: I had no clue that `*` indicated trademarked… I was looking for some footer with an explanation of what it was… So my vote (if I had any) would be to just spell it out `(TM)`….

But as long as we’re hanging on to the asterisk, I think that escaping a metacharacter that is not supposed to be one is essential. And it is flagged by MarkdownLint (I’m sure it is flagged by others as well)?

tomlenth commented:
My apologies.  I'm having a hard time bubbling this up in my priority list and it will likely get worse over the next two weeks. 

We cannot just replace the asterisks with a (TM) since it's not a direct mapping.  it's very odd, but legal has been requiring this for as long as I can remember. 

I also noticed today that we do have some samples using the /* already, but even within the single readme, it isn't consistent.  It does look very odd without being rendered in markdown.  As it is, we do some non-standard formatting to ensure white space that gets rendered away. 

Given that this change would extend across well over a hundred samples, it may be best to get the entire samples workgroup onboard to understand the change.   I'd also like to understand what else we're not doing well before mandating changes across a lot of owners.

dseynhae commented:
I don’t think I introduced `(TM)`, all I meant to do was to escape the `*` that wasn’t supposed to be a metacharacter, so it consistently shows up as an asterisk  amongst all MD rendering applications. I’ll see if I missed one later today.

I do agree that it looks weird in `raw\*` text. Maybe the topic for the bigger audience would be to standardize branding and licensing:

(TM), (R), (C)

No confusion or ambiguity there…

I fixed a few other common linting issues, with the only impact that it renders cleaner in other MD rendering applications, what do I have to do to get this single update approved?


Regards,
***@***.***  ***@***.***>

From: tomlenth ***@***.***>
Sent: Tuesday, November 30, 2021 8:34 PM
To: dseynhae/oneAPI-samples ***@***.***>
Cc: Seynhaeve, Dirk ***@***.***>; State change ***@***.***>
Subject: Re: [dseynhae/oneAPI-samples] MarkdownLimit Proposal (PR #2)


My apologies. I'm having a hard time bubbling this up in my priority list and it will likely get worse over the next two weeks.

We cannot just replace the asterisks with a (TM) since it's not a direct mapping. it's very odd, but legal has been requiring this for as long as I can remember.

I also noticed today that we do have some samples using the /* already, but even within the single readme, it isn't consistent. It does look very odd without being rendered in markdown. As it is, we do some non-standard formatting to ensure white space that gets rendered away.

Given that this change would extend across well over a hundred samples, it may be best to get the entire samples workgroup onboard to understand the change. I'd also like to understand what else we're not doing well before mandating changes across a lot of owners.

—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub<https://github.com/dseynhae/oneAPI-samples/pull/2#issuecomment-983279787>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AM43N62UQTNNTHB4F3J6VV3UOWQSHANCNFSM5ISFK3LA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
